src/celeritas/ext/detail/CelerOpticalPhysics.cc

```diff
@@ -54,6 +54,46 @@ enum class OpticalProcessType
     size_,
 };
 
+/*!
+ * Wrapper around a unique pointer to accomodate keeping track whether we
+ * deleguated ownership to Geant4. We have to assume that Geant4 won't free the
+ * memory before we're done reading it...
+ */
+template<class T>
+class UniquePtrOwner
+{
+  public:
+    explicit UniquePtrOwner(std::unique_ptr<T> ptr) : ptr_(std::move(ptr)) {};
+    UniquePtrOwner(UniquePtrOwner const&) = delete;
+    UniquePtrOwner& operator=(UniquePtrOwner const&) = delete;
+    UniquePtrOwner(UniquePtrOwner&&) noexcept = default;
+    UniquePtrOwner& operator=(UniquePtrOwner&&) noexcept = default;
+
+    T* transferOwnership() noexcept
+    {
+        is_owner_ = false;
+        return ptr_.get();
+    };
+
+    operator T*() const noexcept { return ptr_.get(); };
+
+    T* operator->() const noexcept { return ptr_.get(); };
+
+    ~UniquePtrOwner()
+    {
+        // we transfered ownership, don't free the memory
+        if (!is_owner_)
+        {
+            // NOLINTNEXTLINE(bugprone-unused-return-value)
+            ptr_.release();
+        }
+    };
+
+  private:
+    std::unique_ptr<T> ptr_;
+    bool is_owner_{true};
```

Comment:
```suggestion
template<class T>
class ObservingUniquePtr
{
  public:
    explicit ObservingUniquePtr(std::unique_ptr<T> ptr)
        : uptr_(std::move(ptr)), ptr_{uptr_.get()}
    {}

    CELER_DELETE_COPY_DEFAULT_MOVE(ObservingUniquePtr);

    T* release() noexcept { return uptr_.release(); }
    operator T*() const noexcept { return ptr_; }
    T* operator->() const noexcept { return ptr_; }

  private:
    std::unique_ptr<T> uptr_;
    T* ptr_;
```

I'm guessing by the nomenclature and trailing `;` after the function definitions that you copied this from ATLAS? 😉